sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -481,88 +483,104 @@ case class In(value: Expression, list: Seq[Expression]) extends Predicate {
   override def toString: String = s"$value IN ${list.mkString("(", ",", ")")}"
 
   override def eval(input: InternalRow): Any = {
-    val evaluatedValue = value.eval(input)
-    if (evaluatedValue == null) {
-      null
+    if (list.isEmpty && !legacyNullInEmptyBehavior) {
+      // IN (empty list) is always false under current behavior.
+      // Under legacy behavior it's null if the left side is null, otherwise false (SPARK-44550).
+      false
     } else {
-      var hasNull = false
-      list.foreach { e =>
-        val v = e.eval(input)
-        if (v == null) {
-          hasNull = true
-        } else if (ordering.equiv(v, evaluatedValue)) {
-          return true
-        }
-      }
-      if (hasNull) {
+      val evaluatedValue = value.eval(input)
```

Comment:
Note: most of these changes are just whitespace, use hide whitespace mode for easier review.